### PR TITLE
Created ShellSort in Java

### PR DESCRIPTION
**Pull request by iamvs-2002**
### Have you read the [Contributing Guidelines on Pull Requests](https://github.com/TesseractCoding/NeoAlgo/blob/master/CONTRIBUTING.md#pull-requests)?
Yes, I have read them.
(Write your answer here.)

### Description
Created the implementation of the Shell Sort algorithm in Java. 
(Write your answer here.)

### Checklist

- [x] I've read the contribution guidelines.
- [x] I've checked the issue list before deciding what to submit.
- [x] I've edited the `README.md` and link to my code.

## Related Issues or Pull Requests
#1691 
(Write your answer here.)

### CODE REVIEW

File: Java/sort/ShellSort.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
+class ShellSort
```
Comment (HarshCasper) [inline Java/sort/ShellSort.java]: Add a beginner pitch about your Code. Take a look at the other Code to better understand what I am looking for. It should be just a brief intro about the Algorithm in multi-line comments.

File: Java/sort/ShellSort.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
+class ShellSort 
+{ 
+	static void printArray(int arr[]) 
+    { //function to print the elements of the array
```
Comment (HarshCasper) [inline Java/sort/ShellSort.java]: Inconsistent indentation. Use a formatter to make your indentation consistent.

File: Java/sort/ShellSort.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        //called twice...one before sorting and once after sorting
```
Comment (HarshCasper) [inline Java/sort/ShellSort.java]: Doesn't look grammatically correct to me. Please keep in mind, that the comments are not only for your own reference but also for the reference of others.

File: Java/sort/ShellSort.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+					arr[j] = arr[j - gap]; 
+				arr[j] = temp; 
+			} 
+		} 
+		return 0; 
+	} 
+
+	public static void main(String args[]) 
+	{ 
+        int arr[] = {12, 34, 54, 2, 3}; //this is our array
```
Comment (HarshCasper) [inline Java/sort/ShellSort.java]: We don't prefer to have driver code. Please use Dynamic Input instead of Static Input.

**Review by HarshCasper — CHANGES_REQUESTED** [review]
This Code seems to be plagiarized from this [GeeksforGeeks Article](https://www.geeksforgeeks.org/shellsort/). I will let @atarax665 and @iamrajiv to take a look at it once. 

If found plagiarized, you can take it as a warning. The second warning will count you to be debarred from contributing to the repo forever.

File: Java/sort/ShellSort.java (hunk 1 above)
Comment (iamvs-2002) [reply Java/sort/ShellSort.java]: Ohh okay, I am on it.

**Comment by HarshCasper** [comment]
Hello @iamvs-2002 

I would not be able to merge the Code. 

This Code seems to be plagiarized from this [GeeksforGeeks Article](https://www.geeksforgeeks.org/shellsort/). 

Can you please explain this?

**Comment by iamvs-2002** [comment]
> Hello @iamvs-2002
> 
> I would not be able to merge the Code.
> 
> This Code seems to be plagiarized from this [GeeksforGeeks Article](https://www.geeksforgeeks.org/shellsort/).
> 
> Can you please explain this?

The code is not copied. It's from the algorithm taught in the college class.
The Shell Sort algorithm will have the same flow:
i.e., first, we compare the distant elements and calculate the gap and reduce the gap by 2 after passing through all the elements.

Eg:
array = {63,19,7,90,29,39,58,42,12,2}
n = 10
so initial gap =n/2 = 5
now we will sort the elements at a distance/gap of 5 
so first compare 39 to 63
since 63>39, swap them
array = {39,19,7,90,29,63,58,42,12,2}
now we compare 19 to 58
since they are already sorted, we dont need to change them
now we compare 7 to 42
since they are already sorted, we dont need to change them
now we compare 90 to 12
since 12<90, we swap them
array = {39,19,7,12,29,63,58,42,90,2}
now we compare 29 to 2
since 2<29, we swap them
array = {39,19,7,12,2,63,58,42,90,29}
now we will again calculate the gap = initial gap/2 = 5/2 = 2
now we will compare elements at a distance of 2
array = {39,19,7,12,2,63,58,42,90,29}
compare 39 to 7
since 39>7, swap them
array = {7,19,39,12,2,63,58,42,90,29}
similarly, now 19 and 12 will be swapped
array = {7,12,39,19,2,63,58,42,90,29}
similarly now 39 and 2 will be swapped
array = {7,12,2,19,39,63,58,42,90,29}
19 and 63 dont get swapped
39 and 58 also dont
63 and 42 get swapped
array = {7,12,2,19,39,42,58,63,90,29}
58 and 90 dont get swapped
63 and 29 get swapped
array = {7,12,2,19,39,42,58,29,90,63}
now gap = 2/1 = 1
so we now compare elements at a distance of 1
array = {7,12,2,19,39,42,58,29,90,63}
7 and 12 don't get swapped
12 and 2 do
array = {7,2,12,19,39,42,58,29,90,63}
now 2 will also be compared to 7 since it is less than 7, 2 and 7 will be swapped
array = {2,7,12,19,39,42,58,29,90,63}
now comparing the next ones..
12 and 19 don't get swapped
19 and 39 don't get swapped
39 and 42 don't
42 and 58 don't
58 and 29 do...also 29 will be compared with all the elements of the array till it is larger than some element..
array = {2,7,12,19,29,39,42,58,90,63}
similarly, after a while 90 and 63 get swapped
array = {2,7,12,19,29,39,42,58,63,90}
and now we have the sorted array

**Comment by iamvs-2002** [comment]
@HarshCasper Sir any update for the same?

File: Java/sort/ShellSort.java [hunk 5]
```diff
@@ -0,0 +1,85 @@
+import java.util.Scanner;
+
+/*
+* Shell Sort is a very efficient sorting algorithm based on insertion sort technique.
+* Shell Sort is more effective than insertion sort as
+*	sometimes in Insertion Sorting, we shift a large block to insert an item at the right place as per the sorting order,
+*	which can be avoided by using the Shell Sort algorithm.
+* In Shell Sort, the sorting is done at certain intervals/gaps and everytime, this gap is reduced.
+* This way, Shell Sort becomes nearly twice as fast as Insertion Sort.
```
Comment (HarshCasper) [inline Java/sort/ShellSort.java]: Left-Align it.

**Review by HarshCasper — CHANGES_REQUESTED** [review]
Code looks good. Make the apt README changes as well, and it would be ready to merge.

File: Java/sort/ShellSort.java (hunk 5 above)
Comment (iamvs-2002) [reply Java/sort/ShellSort.java]: Okay sir I am on it.
Please guide what is apt README?

Comment (iamvs-2002) [reply Java/sort/ShellSort.java]: I got it sir. I have update the readme file.

Comment (atarax665) [reply Java/sort/ShellSort.java]: You can update the README.md in this PR itself. Update README.md by visiting ```shellsortinjava``` branch.

Comment (iamvs-2002) [reply Java/sort/ShellSort.java]: Ohh okay

**Comment by iamvs-2002** [comment]
@HarshCasper Done!

**Comment by iamvs-2002** [comment]
@HarshCasper Done!

**Review by iamrajiv — CHANGES_REQUESTED** [review]
@iamvs-2002 Format you code like this

```java
import java.util.Scanner;

/*
*Shell Sort is a very efficient sorting algorithm based on insertion sort technique.
*
*Shell Sort is more effective than insertion sort as
*sometimes in Insertion Sorting, we shift a large block to insert an item at the right place
*as per the sorting order,
*which can be avoided by using the Shell Sort algorithm.
*
*In Shell Sort, the sorting is done at certain intervals/gaps and everytime, this gap is reduced.
*This way, Shell Sort becomes nearly twice as fast as Insertion Sort.
*/

public class ShellSort {
	int shellsort(int arr[], int n) {
		//here n is the size of the array
		//initially gap = n/2
		for (int gap = n / 2; gap > 0; gap = gap / 2) {
			for (int i = gap; i < n; i++) {
				//decreasing the gap
				int k = arr[i];
				int j = i;

				while (j >= gap && arr[j - gap] > k) {
					arr[j] = arr[j - gap];
					j = j - gap;
				}
				arr[j] = k;
			}
		}
		return 0;
	}

	private static Scanner sc = new Scanner(System.in);

	public static void main(String args[]) {
		//we store the number of elements in n
		System.out.println("Kindly enter the number of elements: ");
		int n = sc.nextInt();

		//declaring an array of n elements
		int array[] = new int[n];

		//taking the input from the user
		for (int i = 0; i < n; i++) {
			System.out.print("\n" + i + "'th element : ");
			array[i] = sc.nextInt();
		}

		System.out.println("\nArray before sorting");
		for (int j = 0; j < n; j++)
			System.out.print(array[j] + " ");


		//creating an object of shellsort class so that the methods inside shellsort class can be used
		ShellSort object = new ShellSort();
		object.shellsort(array, n);

		System.out.println("\nArray after sorting");
		for (int k = 0; k < n; k++)
			System.out.print(array[k] + " ");

	}
}

/*
Output:

Kindly enter the number of elements: 5

0'th element : 1
1'th element : 2
2'th element : 5
3'th element : 3
4'th element : 0
Array before sorting
1 2 5 3 0
Array after sorting
0 1 2 3 5
*/
```

**Comment by iamvs-2002** [comment]
I am sorry sir, but I didn't understand. Please guide.

**Comment by iamrajiv** [comment]
@iamvs-2002 Your code was not formatted properly so I formatted the code please use the above code.

**Comment by iamvs-2002** [comment]
> @iamvs-2002 Your code was not formatted properly so I formatted the code please use the above code.

Ohh... thanks for it😊
I have updated the file.

**Review by atarax665 — APPROVED** [review]

**Review by iamrajiv — APPROVED** [review]

**Comment by iamvs-2002** [comment]
@iamrajiv Kindly add the dwoc label as the core team says that the label must be both on the pr and the issue.

**Pull request merged**